tests/test_filtering.py

```diff
@@ -1,333 +1,335 @@
+
+Test Cases:
+- Inclusion and exclusion filters based on OS names, hostnames, and IP addresses.
+- Validation of constraints with regular expressions.
+- Handling of CSV files for filtering, including error handling for missing fields
+  and incorrect file formats.
+- Ensuring that filters work correctly with both IPv4 and IPv6 addresses.
+
+This module also includes helper functions to streamline the creation of filters
+and the writing of CSV files for testing purposes.
+"""
+
 import csv
 
+import pytest  # noqa
 
-def test_filtering_pass_include():
-    """
-    Test the use-case where the constraint is a valid set of "limits"
-    """
-    key_values = [("os_name", "eos"), ("host", ".*nyc1")]
-    constraints = [f"{key}={val}" for key, val in key_values]
-    field_names = [key for key, _ in key_values]
+from netcfgbu.filtering import create_filter
 
-    filter_fn = create_filter(
-        constraints=constraints, field_names=field_names, include=True
-    )
+# Test data
+TEST_DATA1 = {"ipaddr": "10.10.0.2", "host": "switch1.nyc1"}
+TEST_DATA2 = {"ipaddr": "10.10.0.3", "host": "switch1.nyc1"}
+TEST_DATA3 = {"ipaddr": "10.10.0.4", "host": "switch1.dc1"}
+TEST_DATA4 = {"ipaddr": "3001:10:10::2", "host": "switch1.nyc1"}
+TEST_DATA5 = {"ipaddr": "3001:10:10::3", "host": "switch1.nyc1"}
+TEST_DATA6 = {"ipaddr": "3001:10:10::4", "host": "switch1.dc1"}
 
-    assert filter_fn(dict(os_name="eos", host="switch1.nyc1")) is True
-    assert filter_fn(dict(os_name="ios", host="switch1.nyc1")) is False
-    assert filter_fn(dict(os_name="eos", host="switch1.dc1")) is False
 
+def create_and_test_filter(constraints, field_names, include, test_data, expected_results):
+    """Helper function to create a filter and run assertions.
 
-def test_filtering_pass_exclude():
+    Args:
+        constraints (list): List of constraints for the filter.
+        field_names (list): List of field names used in the filter.
+        include (bool): Whether to include or exclude constraints.
+        test_data (list): List of dictionaries representing test data.
+        expected_results (list): List of expected results (True/False).
     """
-    Test use-case where the constraint is a valid set of "excludes"
+    filter_fn = create_filter(constraints=constraints, field_names=field_names, include=include)
+    for test_rec, expected in zip(test_data, expected_results):
+        assert filter_fn(test_rec) is expected
+
+
+def write_csv(tmpfile, fieldnames, rows):
+    """Helper function to write rows to a CSV file.
+
+    Args:
+        tmpfile (Path): Path to the file.
+        fieldnames (list): List of field names for the CSV.
+        rows (list): List of dictionaries representing rows to write.
     """
-    key_values = [("os_name", "eos"), ("host", ".*nyc1")]
-    constraints = [f"{key}={val}" for key, val in key_values]
-    field_names = [key for key, _ in key_values]
+    with open(tmpfile, "w+", encoding="utf-8") as ofile:
```

Comment:
Add more comprehensive error handling for file operations to prevent unhandled exceptions. [medium]